lib/src/license_detection/license.dart

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:meta/meta.dart';
+import 'package:pana/src/third_party/diff_match_patch/diff.dart';
+import 'package:pana/src/license_detection/token_matcher.dart';
+import 'package:pana/src/license_detection/tokenizer.dart';
+
+import 'crc32.dart';
+
+@sealed
+class License {
+  /// SPDX identifier of the license, is empty in case of unknown license.
+  final String identifier;
+
+  /// Original text from the license file.
+  final String content;
+
+  /// Normalized [Token]s created from the original text.
+  final List<Token> tokens;
+
+  /// Map from [Token.value] to their number of occurrences in this license.
+  final Map<String, int> occurrences;
+
+  License._(this.content, this.tokens, this.occurrences, this.identifier);
+
+  factory License.parse(String identifier, String content) {
+    final tokens = tokenize(content);
+    final table = generateFrequencyTable(tokens);
+    return License._(content, tokens, table, identifier);
+  }
+}
+
+String tokensNormalizedValue(Iterable<Token> tokens) {
+  var buffer = StringBuffer();
+
+  for (var token in tokens) {
+    buffer.write(token.value);
+    buffer.write(' ');
+  }
+
+  return buffer.toString().trimRight();
+}
+
+@sealed
+class NGram {
+  /// Text for which the hash value was generated.
+  final String text;
+
+  /// [CRC-32][1] checksum value generated for text.
+  ///
+  /// [1]: https://en.wikipedia.org/wiki/Cyclic_redundancy_check
+  final int checksum;
+
+  /// Index of the first token in the checksum.
+  final int start;
+
+  /// Index of the last token in the checksum.
+  final int end;
+
+  int get granularity => text.split(' ').length;
+
+  NGram(this.text, this.checksum, this.start, this.end);
+}
+
+/// A [License] instance with generated [nGrams].
+// TODO: Change class name to something more meaningful.
+@sealed
+class PossibleLicense {
+  final License license;
+
+  final List<NGram> nGrams;
+
+  final Map<int, List<NGram>> checksumMap;
+
+  PossibleLicense._(this.license, this.nGrams, this.checksumMap);
+
+  factory PossibleLicense.parse(License license, int granularity) {
```

Comment:
Should we consider calling this `n` instead of `granularity`.
```suggestion
  factory PossibleLicense.parse(License license, int n) {
```

Also why don't we store `n` as a property on instances of the class?
I imagine that when we do `findPotentialMatches` it might be wise to have:
```
if (knownLicense.n != unknownLicense.n) {
  throw ArgumentError('n-gram size for knownLicense and unknownLicense must be the same!');
}
```

This way we ensure that we never compare checksums from two different licenses.

Am I correct that comparing licenses with different n-gram sizes would be very bad?

Reply:
Yes it's very bad.